generator/src/main/java/com/graphicsfuzz/generator/mutateapi/ExpressionGenerator.java

```diff
@@ -0,0 +1,298 @@
+import com.graphicsfuzz.common.ast.expr.FunctionCallExpr;
+import com.graphicsfuzz.common.ast.expr.IntConstantExpr;
+import com.graphicsfuzz.common.ast.expr.TypeConstructorExpr;
+import com.graphicsfuzz.common.ast.expr.UIntConstantExpr;
+import com.graphicsfuzz.common.ast.expr.VariableIdentifierExpr;
+import com.graphicsfuzz.common.ast.stmt.BlockStmt;
+import com.graphicsfuzz.common.ast.stmt.DeclarationStmt;
+import com.graphicsfuzz.common.ast.stmt.ReturnStmt;
+import com.graphicsfuzz.common.ast.stmt.Stmt;
+import com.graphicsfuzz.common.ast.type.BasicType;
+import com.graphicsfuzz.common.util.IRandom;
+import com.graphicsfuzz.common.util.IdGenerator;
+import com.graphicsfuzz.common.util.PipelineInfo;
+import com.graphicsfuzz.generator.semanticschanging.LiteralFuzzer;
+import com.graphicsfuzz.util.Constants;
+import com.sun.org.apache.xpath.internal.operations.Bool;
+import java.text.NumberFormat;
+import java.text.ParseException;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.List;
+import java.util.Optional;
+import java.util.stream.Collectors;
+import org.apache.commons.lang3.tuple.ImmutablePair;
+import org.apache.commons.lang3.tuple.Pair;
+
+public class ExpressionGenerator {
+
+  private static final int MAX_FUNCTION_PARAMS = 5;
+  private static final int INT_MIN = -(1 << 17);
+  private static final int INT_MAX = 1 << 17;
+
+  private final TranslationUnit translationUnit;
+  private final PipelineInfo pipelineInfo;
+  private final IdGenerator idGenerator;
+
+  public ExpressionGenerator(TranslationUnit translationUnit, PipelineInfo pipelineInfo) {
+    this.translationUnit = translationUnit;
+    this.pipelineInfo = pipelineInfo;
+    idGenerator = new IdGenerator();
+  }
+
+  private Expr generateLiteralNumber(Value value) {
+    assert BasicType.allScalarTypes().contains(value.getType());
+
+    if (value.getType() == BasicType.FLOAT) {
+      return new FloatConstantExpr(nameFromNumber(value.getData().get(0).get()));
+    }
+    if (value.getType() == BasicType.INT) {
+      return new IntConstantExpr(nameFromNumber(value.getData().get(0).get()));
+    }
+    if (value.getType() == BasicType.UINT) {
+      return new UIntConstantExpr(nameFromNumber(value.getData().get(0).get()) + "u");
+    }
+    List<Expr> args = new ArrayList<>();
+    for (int i = 0; i < value.getType().getNumElements(); i++) {
+      assert value.getType().getElementType() != value.getType();
+      args.add(generateLiteralNumber(
+          new PrimitiveValue(value.getType().getElementType(),
+              Arrays.asList(Optional.of(value.getData().get(i).get()))
+          )));
+    }
+    return new TypeConstructorExpr(value.getType().toString(), args);
+  }
+
+  public Expr generateExpr(FactManager factManager,
+                           FunctionDefinition functionDefinition,
+                           Stmt stmt,
+                           Value value,
+                           IRandom generator) {
```

Comment:
I suggest you structure this as follows:

if (currentDepth >= MAX_DEPTH) {
  return a literal expr
}

currentDepth++;
Expr result = null;
while (true) {
   // the code you have now, but assign to result and break from the loop rather than returning.
}
currentDepth--;
return result;